### PR TITLE
feat(hybrid-cloud): Move Integration/OrgIntegration to Control and implement IntegrationService

### PR DESCRIPTION
**Pull request by leeandher**
Offshoot of https://github.com/getsentry/sentry/pull/42569 (First impl of IdentityService + IntegrationService)
Which is an offshoot of https://github.com/getsentry/sentry/pull/42471 (Impl of both services + slack webhook parser)

Sibling to this PR is https://github.com/getsentry/sentry/pull/42677, which is the Identity Service

---

This PR is scoped down to just the Integration Service and its relevant tests. Even though this is a silo change, since we're adopting the service at key app surfaces like notifications/alerting, it's going to be live in prod, replacing the ORM queries.

### CODE REVIEW

**Comment by corps** [comment]
Working on this PR now

File: src/sentry/services/hybrid_cloud/integration/__init__.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        self, org_integration_id: int, config: Mapping[str, Any], should_clear: bool = False
+    ) -> APIOrganizationIntegration | None:
+        """
+        Returns an APIOrganizationIntegration if the associated org_integration.config value
+        was successfully updated, otherwise returns None.
+        If should_clear is True, runs dict.clear() first, otherwise just uses dict.update()
+        """
+
+    def get_installation(
+        self, integration_id: int, organization_id: int, **kwargs: Mapping[str, Any]
```
Comment (corps) [inline src/sentry/services/hybrid_cloud/integration/__init__.py:150]: This part concerns me, in general having untyped mapping in our protocol could lead to some future grief.  Are we using this?  If we're not using it now, I'd suggest just dropping it for now.

**Review by corps — APPROVED** [review]
My main concern is with the `**kw` of `get_installation`.  I think it's fair to possible just drop it for now?

**Review by vbro — APPROVED** [review]

File: src/sentry/services/hybrid_cloud/integration/impl.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    ) -> ApiPaginationResult:
+        return args.do_hybrid_cloud_pagination(
+            description="page_integration_ids",
+            paginator_cls=OffsetPaginator,
+            order_by="name",
+            queryset=Integration.objects.filter(
+                organizations=organization_id,
+                provider__in=provider_keys,
+            ),
+        )
```
Comment (markstory) [inline src/sentry/services/hybrid_cloud/integration/impl.py:37]: In the RPC call path wouldn't it be more efficient to do the pagination + record fetch as a single operation? The same should be true for the database read path as well.

File: src/sentry/tasks/integrations/sync_assignee_outbound.py [hunk 3]
```diff
@@ -28,13 +29,13 @@ def sync_assignee_outbound(external_issue_id: int, user_id: Optional[int], assig
...
 
-    integration = Integration.objects.get(id=external_issue.integration_id)
     # Assume unassign if None.
     user = user_service.get_user(user_id) if user_id else None
-
-    installation = integration.get_installation(organization_id=external_issue.organization_id)
-
+    installation = integration_service.get_installation(
+        integration_id=integration.id, organization_id=external_issue.organization_id
+    )
```
Comment (markstory) [inline src/sentry/tasks/integrations/sync_assignee_outbound.py:38]: Should we have a service method that combines these queries? There are many workflows in integrations that will need the integration and the organization_integration & installation. Having a service method to get all the context in one pass would help here, and likely in more places.

File: src/sentry/tasks/integrations/sync_status_outbound.py [hunk 4]
```diff
@@ -31,8 +32,10 @@ def sync_status_outbound(group_id: int, external_issue_id: int) -> Optional[bool
         # Issue link could have been deleted while sync job was in the queue.
         return None
 
-    integration = Integration.objects.get(id=external_issue.integration_id)
-    installation = integration.get_installation(organization_id=external_issue.organization_id)
+    integration = integration_service.get_integration(integration_id=external_issue.integration_id)
+    installation = integration_service.get_installation(
+        integration_id=integration.id, organization_id=external_issue.organization_id
+    )
```
Comment (markstory) [inline src/sentry/tasks/integrations/sync_status_outbound.py:38]: Another place where fetching the integration, installation and orgintegration all at once could help.

File: src/sentry/utils/pagination_factory.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    if not cursor_name:
+        return None
+
+    try:
+        return cursor_cls.from_string(cursor_name)
+    except ValueError:
+        raise ValueError("Invalid cursor parameter.")
+
+
+def get_paginatior(
```
Comment (markstory) [inline src/sentry/utils/pagination_factory.py:55]:
```suggestion
def get_paginator(
```

File: src/sentry/services/hybrid_cloud/integration/impl.py (hunk 2 above)
Comment (corps) [reply src/sentry/services/hybrid_cloud/integration/impl.py:37]: My worry, however, is that "to do that as a single operation" requires a fairly complex service interface -- basically supporting all orm filtering as parameterizable.  I think it's ok to take this approach, but let's talk more.

File: src/sentry/services/hybrid_cloud/integration/__init__.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    def get_organization_integration(
+        self, integration_id: int, organization_id: int
+    ) -> APIOrganizationIntegration | None:
+        """
+        Returns an APIOrganizationIntegration from the integration and organization ids.
+        """
+        all_ois = self.get_organization_integrations(integration_id=integration_id)
+        organization_integration = next(
+            (oi for oi in all_ois if oi.organization_id == organization_id), None
+        )
```
[the remark this answers is not in the document]
Comment (corps) [reply src/sentry/services/hybrid_cloud/integration/__init__.py:133]: I also tend to agree that this seems not super useful, esp if it's not an efficient path. In fact, getting all the organizations doing several hand filterings would also be more efficient than making calls to this, which would be fetching the same big set over and over just to filter it down. It's ok if the service interface isn't always convenient. The linux philosophy: prefer simplicity of implementation over simplicity of interface.

File: src/sentry/services/hybrid_cloud/integration/impl.py (hunk 2 above)
Comment (corps) [reply src/sentry/services/hybrid_cloud/integration/impl.py:37]: Paginating back just the ids allows the caller more flexibility and compatibility -- sending each id batch off to, for instance, the serialize method, or using it in compound pagination.

I don't think optimizing for strict performance here is ideal.  The simplest allows the interface to avoid generics, over serialization, and other complexity.  It keeps all options and uses open for a small price.

Also, I think it's absolutely possible to fix the `fetch ids -> serialize ids` at the RPC level using a batch mechanism.  It would be better to solve that wholistically than to complicate service implementations.

File: src/sentry/tasks/integrations/sync_assignee_outbound.py (hunk 3 above)
Comment (corps) [reply src/sentry/tasks/integrations/sync_assignee_outbound.py:38]: That's probably true.  I do like that idea.

File: src/sentry/services/hybrid_cloud/integration/impl.py (hunk 2 above)
Comment (corps) [reply src/sentry/services/hybrid_cloud/integration/impl.py:37]: It's slowly coming back to me again, because I switch contexts alot these days.

So yes, one of the big reasons actually is serializers themselves.  When we're talking about paginating we're talking about implementing API endpoints, and API endpoints tend to involve multiple, and even sometimes conditional, serializers.  Making serialization *part* of the pagination api also means encoding that logic into that pagination aspect of the pagination RPC.  We already have to encode serializer logic into `serialize_Xs`, because we have non paginated use cases or that.  Also encoding it into the pagination protocol risks having to extra complexity since all serialization parameters will then need to be shared between the two.

The main goal here was to make it as easy as possible to add additional pagination RPC endpoints in the future by future developers.  The pattern of "write serialize_Xs and pagination_Xs then combine them" left as few decisions as possible for design for future developers.  It's simple, automatic, fewer moving parts / parameters.

Basically, my intuition is to prefer a simple pattern than can be replicated out with less work.

Comment (corps) [reply src/sentry/services/hybrid_cloud/integration/impl.py:37]: Think of it this way: each endpoint has one, clear, simple role.

`serialize_Xs` => Its one job is to serialize.
`paginate_Xs` => its one job is to query.

File: src/sentry/integrations/slack/webhooks/event.py [hunk 7]
```diff
@@ -134,7 +136,17 @@ def on_link_shared(self, request: Request, slack_request: SlackDMRequest) -> boo
             if link_type is None or args is None:
                 continue
 
-            organization = slack_request.integration.organizations.first()
+            ois = integration_service.get_organization_integrations(
+                integration_id=slack_request.integration.id
+            )
+            organization_id = ois[0].organization_id if len(ois) > 0 else None
```
[the remark this answers is not in the document]
Comment (leeandher) [reply src/sentry/integrations/slack/webhooks/event.py:142]: That makes sense, I'll add it.

File: src/sentry/services/hybrid_cloud/integration/__init__.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (leeandher) [reply src/sentry/services/hybrid_cloud/integration/__init__.py:133]: I have some changes to streamline this, good callout!

File: src/sentry/tasks/integrations/sync_assignee_outbound.py (hunk 3 above)
Comment (leeandher) [reply src/sentry/tasks/integrations/sync_assignee_outbound.py]: Put together the combination service call, but still keeping installation separate since it isn't a serialized RPC call, it returns a python object :+1:

File: src/sentry/api/serializers/models/integration.py [hunk 8]
```diff
@@ -58,7 +65,7 @@ def __init__(
 
     def serialize(
         self,
-        obj: Integration,
+        obj: Integration | APIIntegration,
```
Comment (corps) [inline src/sentry/api/serializers/models/integration.py:68]: This all really, really makes me sad.  Is there no way to avoid serializing APIIntegration objects?

Preferably, serializers always work on ORM objects, and call service methods to serialize related objects that exist in a far off silo.  For instance, `user_service.serialize_users` allows region silos to "serialize" users without making the unfortunate complication of having Serializer classes actually, serialize APIUsers (Which would devolve quickly into pathological ugliness).

Is there, any way we can revert this and say, just implement an integration_service.serialize_integrations call which invokes these in the "aligned" silo?

File: src/sentry/integrations/base.py [hunk 9]
```diff
@@ -308,8 +317,11 @@ def get_client(self) -> Any:
         # Return the api client for a given provider
         raise NotImplementedError
 
-    def get_default_identity(self) -> Identity:
+    def get_default_identity(self) -> Identity | None:
         """For Integrations that rely solely on user auth for authentication."""
+
+        if not self.org_integration:
```
Comment (corps) [inline src/sentry/integrations/base.py:323]: Hmm, did the previous implementation raise `NotFound`?  Should we be instead raising `Identity.NotFound` in this case?

**Review by corps — APPROVED** [review]
I approve with some minor reservations.

I really don't like having Serializer classes serializing hybrid cloud data classes, in general, because I think if we start doing that too much we end up with really ambiguous weird typing and a ton of extra service calls.  Ideally, serializers are run in the silo most aligned to the model.  Preferably, we'd have a service method to do serialization that then calls the serializer from ids in hand.  

I also have slight concern about the switching to returning None instead of the NotFound exception, but I am not confident that this is an actual issue, it would depend on the callsites and behavior of the original method which I am not familiar.

That said, this is a big, important PR that gets a hugely important feature "working" with tests in hybrid cloud.  And changing the serializer later, or unwinding that, isn't the end of the world.  If you can validate that this PR works with the ecosystems team and doesn't break our existing integrations, and get it merged + deployed, I'd call it a win.  We can address finer details in follow up.

I live this your call.  Let me know if you need my help deploying this safely.

File: src/sentry/api/serializers/models/integration.py (hunk 8 above)
Comment (leeandher) [reply src/sentry/api/serializers/models/integration.py:68]: @corps  Encountering some issues with this since the serializers themselves can't safely be called by the Control Silo (they use Region Silo models in `get_attrs`. Do you have any ideas for how to avoid this? 

Context: https://github.com/getsentry/sentry/blob/master/src/sentry/api/serializers/models/integration.py#L202-L214

I can't have the service accept the `Group` instance, so I have to pass an ID, but the serializer expects the actual `Group` instance, which is not available in the control silo.

File: src/sentry/integrations/base.py (hunk 9 above)
Comment (leeandher) [reply src/sentry/integrations/base.py:322]: Good call, fixed

File: src/sentry/api/serializers/models/integration.py (hunk 8 above)
Comment (leeandher) [reply src/sentry/api/serializers/models/integration.py]: Addressed, pinned the type at APIIntegration and moved the serializers.

**Comment by asottile-sentry** [comment]
paged us: this breaks getsentry so I'm reverting
